pkg/lib/cockpit-upload-helper.js

```diff
@@ -0,0 +1,164 @@
+ * optional progress callback using the fsreplace1 channel.
+ *
+ * Example usage:
+ *
+ * const helper = UploadHelper("/tmp/file.txt")
+ * try {
+ *   const status = await helper.upload(file);
+ * } catch (exc) {
+ * }
+ *
+ * Uploads can be cancelled while in progress by calling `helper.cancel`,
+ * fsreplace1 will automatically clean up the temporary file.
+ */
+export class UploadHelper {
+    constructor(destination, onProgress, options) {
+        this.destination = destination;
+        this.progressCallback = onProgress;
+        this.outstanding = 0;
+
+        // Promise resolvers
+        this.resolveHandler = null;
+
+        options = options || {};
+        this.channel = cockpit.channel({
+            binary: true,
+            payload: "fsreplace1",
+            path: this.destination,
+            "send-acks": "frames",
+            ...options,
+        });
+        this.channel.addEventListener("control", this.on_control.bind(this));
+    }
+
+    // Private methods
+    on_control(event, message) {
+        if (message?.command === "ack") {
+            this.outstanding -= message.frames;
+
+            console.assert(this.outstanding >= 0, "outstanding went negative", this.outstanding);
+
+            if (this.resolveHandler) {
+                this.resolveHandler();
+                this.resolveHandler = null;
+            }
+        }
+    }
+
+    read_chunk(blob) {
+        return new Promise((resolve, reject) => {
+            const reader = new FileReader();
+            reader.onload = (event) => resolve(event.target.result);
+            reader.onerror = reject;
+            reader.readAsArrayBuffer(blob);
+        });
+    }
+
+    write(chunk) {
+        console.assert(this.outstanding >= 0, "ack outstanding went negative", this.outstanding);
+        this.outstanding += 1;
+        this.channel.send(chunk);
+
+        if (this.outstanding < TOTAL_FRAME_WINDOW) {
+            return Promise.resolve();
+        } else {
+            return new Promise((resolve, reject) => {
+                this.resolveHandler = resolve;
+            });
+        }
+    }
+
+    close() {
+        const channel = this.channel;
+        return new Promise((resolve, reject) => {
+            channel.addEventListener("close", function(event, message) {
+                if (message.problem) {
+                    reject(message);
+                } else {
+                    resolve();
+                }
+            });
+            channel.control({ command: "done" });
+        });
+    }
+
+    wait_ready() {
+        const channel = this.channel;
+        return new Promise((resolve, reject) => {
+            channel.addEventListener("ready", function(event, message) {
+                resolve();
+            });
+        });
+    }
+
+    // Public methods
+
+    // Flush is waiting on acks in flight
+    async upload(file) {
+        let close_message = null;
+        this.channel.addEventListener("close", (event, message) => {
```

Comment:
It would be nice if there was a unified top-level handling of close messages, as with 'ack'.  The way that problem codes are dealt with is also duplicated (and different in each case).

Reply:
In fact, I feel like we get little value from upload() and close() being separate functions...

Reply:
Mwoah, not entirely. If we merged the promises we would end up having one eventlistener with complexity